src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+
+			igraph_vector_int_append(vectorlist_all, &vectorlist_1);
+			igraph_vector_int_append(vectorlist_all, &vectorlist_2);
+
+			igraph_vector_int_append(edgelist_all, &edgelist_1);
+			igraph_vector_int_append(edgelist_all, &edgelist_2);
+
+			igraph_vector_int_destroy(&vectorlist_2);
+			igraph_vector_int_destroy(&vectorlist_1);
+
+			igraph_vector_int_destroy(&edgelist_1);
+			igraph_vector_int_destroy(&edgelist_2);
+
+			IGRAPH_FINALLY_CLEAN(4);
+
+			std::set<igraph_integer_t> min_F;
+			min_F.insert(F1);
+		}
+
+		m = k;
+		D = min_F;
+
+		igraph_vector_int_destroy(&vectorlist);
+		igraph_vector_int_destroy(&edgelist);
+		IGRAPH_FINALLY_CLEAN(2);
+	}
+
+	return IGRAPH_SUCCESS;
+}
+
+/* \function  igraph_steiner_dreyfus_wagner
+ *  \brief This function calculates Steiner Tree on undirected graph using Dreyfus-Wagner algorithm
+ *
+ *
+ *	</para><para>
+ *   Steiner Tree is a tree with minimum length connecting steiner terminals which are subset of vertices
+ *   in a graph. If Steiner terminals equal to number of vertices in the graph then problem is reduced
+ *   to minimum spanning tree.
+ *	The Steiner Tree problem is NP-Hard but since it's FPT, Dreyfus-Wagner algorithm is able to calculate
+ *	Exact value of Steiner Tree on smaller graphs.
+ *
+ *	The algorithm uses Dynamic Programming approach and it assumes the graph is undirected.
+ *
+ * </para><para>
+ * Reference:
+ * S.E Dreyfus, R.A Wagner
+ * The steiner problem in graphs
+ * Networks Journal, 1971
+ * https://doi.org/10.1002/net.3230010302
+ *
+ * \param graph The graph object.
+ *
+ * \param res Pointer to a real number, this will contain the result which is distance of Steiner Tree.
+ *
+ * \param res_tree Pointer to a vector, this will contain the Edges that are part of Steiner Tree
+ *
+ * \param steiner_terminals Pointer to a vector, this will contain vertices
+ *
+ * \param mode How to determine the local neighborhood of each vertex
+ *  in directed graphs. Ignored in undirected graphs.
+ * \clist
+ *         \cli IGRAPH_ALL
+ *              take both in- and out-neighbours;
+ *              this is a reasonable default for high-level interfaces.
+ *         \cli IGRAPH_OUT
+ *              take only out-neighbours
+ *         \cli IGRAPH_IN
+ *              take only in-neighbours
+ *         \endclist
+ *
+ * \param weights The edge weights. All edge weights must be
+ *       non-negative. Additionally, no edge weight may be NaN.
+ *
+ * \return Error code:IGRAPH_ERROR
+ *
+ * Time complexity: O( 3^k ∗ V + 2^k ∗V^2 + V∗(V+E) ∗ log(V) )
+ * where V is vertices and E is edges in the graph and k is set of steiner terminals.
+ * It's recommended that V <= 50 and k < 11
+ *
+ * \sa \ref igraph_distances_johnson(), generateSubsets(), fetchIndexofMapofSets(), generate_steiner_tree_appx()
+ */
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph, const igraph_vector_int_t *steiner_terminals,
```

Comment:
Include `core/exceptions.h` and add `IGRAPH_HANDLE_EXCEPTIONS_BEGIN` and `IGRAPH_HANDLE_EXCEPTIONS_END` to the start and end of this function. This will make sure that C++ exceptions cannot escape into C code.